dart/common/Subscription.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2014-2015, Georgia Tech Research Corporation
+ * All rights reserved.
+ *
+ * Author(s): Michael X. Grey <mxgrey@gatech.edu>
+ *
+ * Georgia Tech Graphics Lab and Humanoid Robotics Lab
+ *
+ * Directed by Prof. C. Karen Liu and Prof. Mike Stilman
+ * <karenliu@cc.gatech.edu> <mstilman@cc.gatech.edu>
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_SUBSCRIPTION_H_
+#define DART_COMMON_SUBSCRIPTION_H_
+
+#include <set>
+
+namespace dart {
+namespace common {
+
+class Subscriber;
+
+class Subscription
+{
+public:
+
+  friend class Subscriber;
+
+  /// Destructor will notify all Subscribers that it is destructing
+  virtual ~Subscription();
+
+protected:
+
+  /// Send a notification to all Subscribers
+  void sendNotification(int _notice) const;
```

Comment:
I thought about it, but there are a couple reasons I decided to only offer enumerated notifications:

1. The Subscription concept isn't meant for heavy duty message passing, nor is it meant for robust or general purpose data transfer. There are already many tools for that, like Qt's signals and slots. The notifications are just meant to be a light-weight tool for informing interested parties when they should update.

2. I think an enumerated flag and a pointer to the object that sent it should be sufficient for subscribers to obtain the data that they need from that object. A use case that demands more than that would be better off using a more robust and comprehensive pipeline.